batched-tx-disperse-multisend-py/.github/README.md

```diff
@@ -0,0 +1,98 @@
+
+## Description
+
+Disperse / Multisend are apps used to distribute ERC-20 to multiple addresses in one transaction.
+Although they are useful to reduce gas fees by sending transactions in batches, many use them as part of their scam schemes. 
+
+The goal of this bot is to alert when someone is using any of these apps in order to send native tokens or any other ERC-20 token.
+
+## Supported Chains
+
+The bot is specific to the contracts [`Disperse`][etherscan-contract-disperse] and [`Multisend`][etherscan-contract-multisend].
+
+So it runs on a single chain:
+
+- Ethereum
+
+## Alerts
+
+The bot only emits `info` alerts:
+
+- `BATCHED-ERC20-TX`:
+  - Fired when a transaction calls either:
+    - `disperseToken`, `disperseTokenSimple` from the [`Disperse` contract][etherscan-contract-disperse]
+    - or `multisendToken` from the [`Multisend` contract][etherscan-contract-multisend]
+  - Severity is "low" in case of a manual call, otherwise it is set to "info" (call via the web app)
+  - Type is always set to "info"
+  - Labels:
+    - the origin address of the transaction
+  - Metadata:
+    - `transactions`: the serialized list of arguments for each transaction, IE a list of `(recipient, value)` tuples
+    - `count`: the number of transfers contained in the batch
+- `BATCHED-ETH-TX`:
+  - Fired when a transaction calls either:
+    - `disperseEther` from the [`Disperse` contract][etherscan-contract-disperse]
+    - or `multisendEther` from the [`Multisend` contract][etherscan-contract-multisend]
+  - Severity is always set to "info"
+  - Type is always set to "info"
+  - Labels:
+    - the origin address of the transaction
+  - Metadata:
+    - `transactions`: the serialized list of arguments for each transaction, IE a list of `(recipient, value)` tuples
+    - `count`: the number of transfers contained in the batch
+
+## Configuration
+
+The file [`constants.py`](src/constants.py) contains filtering options.
+
+- `TOKEN`:
+  - should be either the empty string `''` or an address like `'0x767fe9edc9e0df98e07454847909b5e959d7ca0e'`
+  - if empty the agent reports all findings
+  - otherwise it will only report transfers batched for a specific ERC20 token
+
+## Deployment
+
+The code is bundled in a Docker container.
+
+## Tests
+
+The test can be run with `python -m pytest`.
+
+### Data
+
+The agent behaviour can be verified on the following transactions:
+
+- ETH transactions:
+  - Disperse: [0xa7f0f0470e9be92b10c57273087cef31774c1284acf3d3b56e3e92c504437fb4][etherscan-tx-disperse-eth]
+  - Multisend: [0xad2f3e0c98a8951214de9bf0aa1a46684e8f594ceec2e34a81eaa24637cfe771][etherscan-tx-multisend-eth]
+- Token transactions:
+  - Disperse: [0x2e311b6e9c842e4ec06712cad2acb6be9d6eec341c348a7dc3aac51ec9a8426c][etherscan-tx-disperse-token]
+  - Multisend: [0x78b093c64e09cb7a3ce6bad2480549b058550faa5ba21be7c19ad732dc761fc5][etherscan-tx-multisend-token]
+
+## Metrics
+
+The bot looks for transactions to the following contract:
+
+- Ethereum:
```

Comment:
I think this works, but it is very specific to these contracts. If we take a step back, is there a generic way to do this. 

1. As part of the transaction input, we get a set of addresses that receive funds along with the source of the funds.
2. for ERC-20, we see transfer log events, which should allow us to discern whether a distribution of funds happened. 
3. for native, its a bit more tricky as we can only see the transfers with tracing enabled. However, if one reviews the balance of the accounts prior vs after this particular block, the balance should have increased.

This would allow to create this bot in a generic way. WDYT?

Reply:
Yes 100%!
I actually resisted these ideas and went for the naive but functional solution first.
I wondered if 3 was possible ! You mean tracing as in debug trace??
So native transfers don't trigger native events or sthg by chance?